nixos/modules/services/web-apps/privatebin.nix

```diff
@@ -0,0 +1,228 @@
+      example = lib.literalExpression ''
+        {
+          main = {
+            name = "NixOS Based Privatebin";
+            discussion = false;
+            defaultformatter = "plalib.types.intext";
+            qrcode = true
+          };
+          model.class = "Filesystem";
+          model_options.dir = "/var/lib/privatebin/data";
+        }
+      '';
+      type = lib.types.submodule { freeformType = lib.types.attrsOf lib.types.anything; };
+    };
+  };
+
+  config = lib.mkIf cfg.enable {
+
+    services.privatebin.settings = {
+      main = lib.mkDefault { };
+      model.class = lib.mkDefault "Filesystem";
+      model_options.dir = lib.mkDefault "${cfg.dataDir}/data";
+      purge.dir = lib.mkDefault "${cfg.dataDir}/purge";
+      traffic = {
+        dir = lib.mkDefault "${cfg.dataDir}/traffic";
+        header = "X_FORWARDED_FOR";
+      };
+    };
+
+    services.phpfpm.pools.privatebin = {
+      inherit user group;
+      phpPackage = pkgs.php83;
+      phpOptions = ''
+        log_errors = on
+      '';
+      settings = {
+        "listen.mode" = lib.mkDefault "0660";
+        "listen.owner" = lib.mkDefault user;
+        "listen.group" = lib.mkDefault group;
+        "pm" = lib.mkDefault "dynamic";
+        "pm.max_children" = lib.mkDefault 32;
+        "pm.start_servers" = lib.mkDefault 2;
+        "pm.min_spare_servers" = lib.mkDefault 2;
+        "pm.max_spare_servers" = lib.mkDefault 4;
+        "pm.max_requests" = lib.mkDefault 500;
+      };
+      phpEnv.CONFIG_PATH = lib.strings.removeSuffix "/conf.php" (builtins.toString privatebinSettings);
+    };
+
+    services.nginx = lib.mkIf cfg.enableNginx {
+      enable = true;
+      recommendedTlsSettings = lib.mkDefault true;
+      recommendedOptimisation = lib.mkDefault true;
+      recommendedGzipSettings = lib.mkDefault true;
+      virtualHosts.${cfg.virtualHost} = {
+        root = "${cfg.package}";
+        locations = {
+          "/" = {
+            tryFiles = "$uri $uri/ /index.php?$query_str";
+            index = "index.php";
+            extraConfig = ''
+              sendfile off;
+            '';
+          };
+          "~ \.php$" = {
+            extraConfig = ''
+              include ${config.services.nginx.package}/conf/fastcgi_params ;
+              fastcgi_param SCRIPT_FILENAME $request_filename;
+              fastcgi_param modHeadersAvailable true; #Avoid sending the security headers twice
+              fastcgi_pass unix:${config.services.phpfpm.pools.privatebin.socket};
+            '';
+          };
+        };
+      };
+    };
+
+    systemd.tmpfiles.settings."10-privatebin" =
+      lib.attrsets.genAttrs
+        [
+          "${cfg.dataDir}/data"
+          "${cfg.dataDir}/traffic"
+          "${cfg.dataDir}/purge"
+        ]
+        (n: {
+          d = {
+            group = group;
+            mode = "0750";
+            user = user;
+          };
+        });
+
+    users = {
+      users = lib.mkIf (user == defaultUser) {
```

Comment:
Is it possible to use Systemd's `dynamicUser` directive instead?

Reply:
Good question. Let me test it out.

I'm personally not in favour, though, because I think it makes the configuration less declarative. If I have root on `tmpfs` and I rely on uids to keep my users/groups aligned with filesystem permissions, I will have to preserve the state of `/etc/passwd`, `/etc/group` and `/etc/shadow` for `DynamicUser` to work properly.

Reply:
Okay, I was wrong. As per [the documentation](https://www.freedesktop.org/software/systemd/man/latest/systemd.exec.html#DynamicUser=), it seems to be managed independently.

Reply:
Okay, so it doesn't work because php-fpm has to run as root currently.

Reply:
@drupol Does this answer your question?

Reply:
> Okay, so it doesn't work because php-fpm has to run as root currently.

So we have a php-fpm module. Is it possible to use that instead?

Reply:
We are using that @Aleksanaa. It's a limitation of the current `php-fpm` module. You can have a look at how it's defined [here](https://github.com/NixOS/nixpkgs/blob/759537f06e6999e141588ff1c9be7f3a5c060106/nixos/modules/services/web-servers/phpfpm/default.nix#L251C5-L251C32)